### PR TITLE
feat(content-explorer): refactor breadcrumbs and allow props config

### PR DESCRIPTION
**Pull request by ivanthai**
### Overview
`Breadcrumbs` was lacking flexibility to configure certain behaviors. These changes were needed to add responsive behaviors.

### Changes
The new changes to `Breadcrumbs` are
- Custom icon
- Number of items allowed before overflow (default is 1)

`ContentExplorerModal`
- inner components take in a `breadcrumbsProp` attribute to allow EUA to customize it.
- if isResponsive={true}, hide the "back" button ("<")

### Other Approaches
For `ContentExplorerModal`, I tried allowing the passing of a custom `Breadcrumbs` component to completely override the default one. The issue with that is that the breadcrumbs in EUA is not compatible and extensive refactoring is required to allow that. I've settled with refactoring `ItemListBreadcrumbs` in EUA to use `Breadcrumbs` on a separate note.

### Notes
No tests have been written yet and will be added after the approach has been approved

### Demo (Move/Copy Modal in EUA)

breadcrumbs-update

### CODE REVIEW

File: src/features/content-explorer/content-explorer-modal-container/ContentExplorerModalContainer.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ import NewFolderModal from '../new-folder-modal';
 
 class ContentExplorerModalContainer extends Component {
     static propTypes = {
+        /** Breadcrumb component options */
+        breadcrumbProps: PropTypes.object,
```
Comment (greathmaster) [inline src/features/content-explorer/content-explorer-modal-container/ContentExplorerModalContainer.js:12]: Can we create a type for this? Or perhaps define the prop types in the object instead of just `object`?

File: src/features/content-explorer/content-explorer-modal/ContentExplorerModal.js [hunk 2]
```diff
@@ -9,6 +9,7 @@ import ContentExplorer from '../content-explorer';
 import './ContentExplorerModal.scss';
 
 type Props = {
+    breadcrumbProps?: Object,
```
Comment (greathmaster) [inline src/features/content-explorer/content-explorer-modal/ContentExplorerModal.js:12]: Similar to the above...can we be more specific than `Object`?

**Review by greathmaster — COMMENTED** [review]
A couple of general things I noticed

The icon for All Files still shows in the overflow dropdown menu, and also the order of items in the dropdown...should reversed? With `All Files` at the top?
Screen Shot 2022-08-22 at 3 43 27 PM

I also noticed hovering on the menu items makes the menu expand and contract.

Aug-22-2022 16-40-22

Also I think we need to add a grey border to the Folder Tree button.

**Comment by ivanthai** [comment]
Interestingly, that UI issue with hover is present in prod

**Comment by ivanthai** [comment]
ContentExplorer's breadcrumb hover issues have been fixed

content-explorer-breadcrumbs-hover-fix

**Comment by ivanthai** [comment]
FolderTreeIcon styling will be handled in EUA

**Comment by ivanthai** [comment]
FYI, the Dropdown menu button states (e.g. selected/focused) will be handled separately. All Dropdown menu implementations have this problem

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js [hunk 3]
```diff
@@ -28,23 +28,25 @@ const ContentExplorerBreadcrumbs = ({
         >
             <IconChevron direction="left" size="6px" color="#333" />
         </Button>
-        <Breadcrumb label={formatMessage(messages.breadcrumb)}>
+        <Breadcrumb label={formatMessage(messages.breadcrumb)} {...breadcrumbProps}>
             {foldersPath.map((folder, i) => (
-                <PlainButton
+                <div
```
Comment (greathmaster) [inline src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:33]: Just wanted to confirm...does this affect A11Y? Since it's going from a `button` rendered by `<PlainButton>` to a `div`?

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js [hunk 4]
```diff
@@ -28,23 +28,25 @@ const ContentExplorerBreadcrumbs = ({
...
                     className="lnk"
                     key={folder.id}
-                    title={folder.name}
                     onClick={event => onBreadcrumbClick(i, event)}
+                    role="presentation"
```
Comment (greathmaster) [inline src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:37]: If we are updating the above to a `div` the aria role might have to be set to `button`.

File: src/features/content-explorer/content-explorer/ContentExplorerHeaderActions.js [hunk 5]
```diff
@@ -15,6 +15,7 @@ const isSearchResultsFolder = folder => folder && folder.id === SEARCH_RESULTS_F
 
 class ContentExplorerHeaderActions extends Component {
     static propTypes = {
+        breadcrumbProps: PropTypes.object,
```
Comment (greathmaster) [inline src/features/content-explorer/content-explorer/ContentExplorerHeaderActions.js:18]: Do we need to use the Breadcrumb PropType from below here?

File: src/features/content-explorer/content-explorer-modal/ContentExplorerModal.js [hunk 6]
```diff
@@ -2,13 +2,15 @@
 import * as React from 'react';
 import classNames from 'classnames';
 
+import type { BreadcrumbProps } from '../../../components/breadcrumb/Breadcrumb';
```
Comment (tjuanitas) [inline src/features/content-explorer/content-explorer-modal/ContentExplorerModal.js:5]: I think types are usually imported after components. We'll look into adding this as a prettier configuration.

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js (hunk 3 above)
Comment (tjuanitas) [inline src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:33]: What's the reason for this change? Is it style related?

File: src/features/content-explorer/prop-types.js [hunk 7]
```diff
@@ -43,7 +43,15 @@ const ItemsPropType = PropTypes.arrayOf(ItemOrPlaceholderPropType);
 
 const ItemsMapPropType = PropTypes.objectOf(ItemPropType);
 
+const BreadcrumbPropType = PropTypes.shape({
+    itemsBeforeOverflow: PropTypes.number,
+    overflowIcon: PropTypes.node,
+    threshold: PropTypes.number,
+    className: PropTypes.string,
```
Comment (tjuanitas) [inline src/features/content-explorer/prop-types.js:50]: nit: could we sort this?

File: src/features/content-explorer/prop-types.js [hunk 8]
```diff
@@ -43,7 +43,15 @@ const ItemsPropType = PropTypes.arrayOf(ItemOrPlaceholderPropType);
 
 const ItemsMapPropType = PropTypes.objectOf(ItemPropType);
 
+const BreadcrumbPropType = PropTypes.shape({
```
Comment (tjuanitas) [inline src/features/content-explorer/prop-types.js:46]: Does it make sense to use the type from the `Breadcrumb` component instead of creating this new type?
https://github.com/box/box-ui-elements/blob/6c1566e82a3083d46aa6939307d29dd6daf7c3f5/src/components/breadcrumb/Breadcrumb.js#L14-L27

File: src/components/breadcrumb/EllipsisCrumb.js [hunk 9]
```diff
@@ -7,11 +7,12 @@ import PlainButton from '../plain-button';
 
 type Props = {
     children: React.Node,
+    icon?: React.Node | string | null,
```
Comment (tjuanitas) [inline src/components/breadcrumb/EllipsisCrumb.js:10]: When is icon `null`?

File: src/components/breadcrumb/EllipsisCrumb.js [hunk 10]
```diff
@@ -7,11 +7,12 @@ import PlainButton from '../plain-button';
...
 };
 
-const EllipsisCrumb = ({ children }: Props) => (
+const EllipsisCrumb = ({ children, icon }: Props) => (
     <DropdownMenu>
-        <PlainButton className="breadcrumb-toggler">⋯</PlainButton>
+        <PlainButton className="breadcrumb-toggler">{icon}</PlainButton>
```
Comment (tjuanitas) [inline src/components/breadcrumb/EllipsisCrumb.js:15]: Should we default to `...` when `icon` is undefined?

File: src/features/content-explorer/content-explorer-modal-container/ContentExplorerModalContainer.js [hunk 11]
```diff
@@ -2,12 +2,14 @@ import PropTypes from 'prop-types';
 import React, { Component } from 'react';
 import classNames from 'classnames';
 
-import { ContentExplorerModePropType, FoldersPathPropType, ItemsPropType } from '../prop-types';
+import { BreadcrumbPropType, ContentExplorerModePropType, FoldersPathPropType, ItemsPropType } from '../prop-types';
```
Comment (tjuanitas) [inline src/features/content-explorer/content-explorer-modal-container/ContentExplorerModalContainer.js:5]: Should this use the other `BreadcrumbProps` type?

File: src/components/breadcrumb/EllipsisCrumb.js (hunk 9 above)
Comment (tjuanitas) [inline src/components/breadcrumb/EllipsisCrumb.js:10]: I think `string` and `null` are included in `React.Node`

File: src/components/breadcrumb/Breadcrumb.js [hunk 12]
```diff
@@ -10,53 +11,75 @@ import messages from './messages';
...
-        );
-    });
-};
-
-type Props = {
+export type BreadcrumbProps = {
     /** Class name for Breadcrumb */
     children: React.Node,
     /** Breadcrumb contents */
     className?: string,
```
Comment (tjuanitas) [inline src/components/breadcrumb/Breadcrumb.js:18]: Could you swap the descriptions for these props?

File: src/components/breadcrumb/Breadcrumb.js [hunk 13]
```diff
@@ -10,53 +11,75 @@ import messages from './messages';
...
+    /** Show number of breadcrumb items before overflow. Default is 1 */
+    itemsBeforeOverflow?: number,
```
Comment (tjuanitas) [inline src/components/breadcrumb/Breadcrumb.js:20]: Is this the number of crumb items that are displayed before the Ellipsis crumb?

I.e. in this screenshot, does `items` refer to the house icon?
Screen Shot 2022-08-31 at 6 56 50 PM

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js (hunk 3 above)
Comment (ivanthai) [reply src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:33]: Yes, PlainButton does not play nice with Breadcrumbs

File: src/components/breadcrumb/Breadcrumb.js (hunk 13 above)
Comment (ivanthai) [reply src/components/breadcrumb/Breadcrumb.js:20]: Yes, the house icon is the number of items

File: src/components/breadcrumb/EllipsisCrumb.js (hunk 9 above)
Comment (ivanthai) [reply src/components/breadcrumb/EllipsisCrumb.js:10]: When a custom icon is not passed in

File: src/features/content-explorer/prop-types.js (hunk 8 above)
Comment (ivanthai) [reply src/features/content-explorer/prop-types.js:46]: React types and TS types are different and not interchangeable from what I understand

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js (hunk 3 above)
Comment (ivanthai) [reply src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:33]: I changed it so it doesn't. We still use a Button, but applied styling to a div wrapper

File: src/components/breadcrumb/EllipsisCrumb.js [hunk 14]
```diff
@@ -7,13 +7,17 @@ import PlainButton from '../plain-button';
...
+    menuButton?: React.Node,
 };
 
-const EllipsisCrumb = ({ children }: Props) => (
-    <DropdownMenu>
-        <PlainButton className="breadcrumb-toggler">⋯</PlainButton>
-        <Menu>{children}</Menu>
-    </DropdownMenu>
-);
+const EllipsisCrumb = ({ children, menuButton }: Props) => {
```
Comment (ivanthai) [inline src/components/breadcrumb/EllipsisCrumb.js:13]: This used to be an icon, but now is a button to consider custom styling and restrict one specific component rather than multiple

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js [hunk 15]
```diff
@@ -28,23 +29,25 @@ const ContentExplorerBreadcrumbs = ({
...
-                <PlainButton
-                    className="lnk"
-                    key={folder.id}
-                    title={folder.name}
-                    onClick={event => onBreadcrumbClick(i, event)}
-                >
-                    {i === 0 && <IconAllFiles />}
-                    <span>{folder.name}</span>
-                </PlainButton>
+                <div key={folder.id} className="lnk">
```
Comment (ivanthai) [inline src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:34]: I previously changed this from PlainButton -> div, but to preserve a11y and styling, I opted to wrap the PlainButton within a div.

**Review by greathmaster — DISMISSED** [review]
I think this is okay.

File: src/components/breadcrumb/Breadcrumb.js (hunk 13 above)
Comment (tjuanitas) [reply src/components/breadcrumb/Breadcrumb.js:20]: Is there a way we could name this so that it's more obvious this is a number? Without reading the description, it sounds like it could be a list of components

Comment (tjuanitas) [reply src/components/breadcrumb/Breadcrumb.js:20]: maybe something like `[...]Count`?

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (ivanthai) [reply src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:34]: It conflicts with PlainButton's styling

File: src/components/breadcrumb/Breadcrumb.js (hunk 13 above)
Comment (ivanthai) [reply src/components/breadcrumb/Breadcrumb.js:20]: i renamed numItemsBeforeOverflow

File: src/components/breadcrumb/Breadcrumb.js [hunk 16]
```diff
@@ -10,53 +11,75 @@ import messages from './messages';
...
+        if (length > threshold) {
+            newChildren = constructChildren();
+            length = newChildren.length;
+            hasEllipsis = true;
+        }
+
+        return React.Children.map(newChildren, (item, i) => {
+            const isLastCrumb = length === 0 || i === length - 1;
+            return (
+                <Crumb className={hasEllipsis && i === 1 ? 'no-shrink' : undefined} isLastCrumb={isLastCrumb}>
```
[the remark this answers is not in the document]
Comment (ivanthai) [reply src/components/breadcrumb/Breadcrumb.js:70]: Don't think so. Classname is for `<nav>`, not its nested list `Crumbs` is nested within the `nav` element

[the remark this answers is not in the document]
Comment (tjuanitas) [reply src/components/breadcrumb/Breadcrumb.js:70]: I meant the classNames() function to simplify `hasEllipsis && i === 1 ? 'no-shrink' : undefined`

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js [hunk 17]
```diff
@@ -28,23 +29,25 @@ const ContentExplorerBreadcrumbs = ({
...
+                    <PlainButton
+                        className="breadcrumb-lnk"
```
Comment (tjuanitas) [inline src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js:36]: Do we need this class? Is it doing anything?

File: src/components/breadcrumb/Breadcrumb.js [hunk 18]
```diff
@@ -10,53 +11,75 @@ import messages from './messages';
...
-);
+type Props = BreadcrumbProps & InjectIntlProvidedProps;
+
+const Breadcrumb = ({
+    children,
+    className = '',
+    intl,
+    numItemsBeforeOverflow = 1,
+    overflowMenuButton,
+    hasReverseOverflowOrder = false,
```
Comment (tjuanitas) [inline src/components/breadcrumb/Breadcrumb.js:37]: could you move this before `intl`?

File: src/components/breadcrumb/Breadcrumb.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (ivanthai) [reply src/components/breadcrumb/Breadcrumb.js:70]: Ah, brain wasn't screwed on right that day. That's a good idea

File: src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js (hunk 17 above)
Comment (ivanthai) [reply src/features/content-explorer/content-explorer/ContentExplorerBreadcrumbs.js]: I changed this to a test-id attr since it is only used within a testing context

**Review by tjuanitas — APPROVED** [review]
